### PR TITLE
Perspective projection and visualization of world/ego points in Carla

### PR DESCRIPTION
**Pull request by Haichao-Zhang**
Added support to project and visualize points in world and ego coordinates in the camera image.
This is useful to get a better understanding of items such as waypoints, trajectories etc. through visualization.

This type of rendering is different from ```debug.draw_point()``` function from Carla in the following aspects:
1. ```debug.draw_point()``` will paint the point in the world, which can modify the sensor perception. This is not desirable and has side-effects in some cases, e.g. it will impact the input to a learned model.
2. ```debug.draw_point()``` has one time-step lag. The points only become visible in the next time step. This is not desirable when we want to debug items that are changing at each time step.

An example usage for visualizing the waypoint (with interpolation) is shown here:
ezgif com-gif-maker (10)

### CODE REVIEW

File: alf/environments/carla_sensors.py [hunk 1]
```diff
@@ -616,10 +618,124 @@ def render(self, display):
...
+        point_in_camera_coords = point_in_camera_coords[:, valid_ind]
+
+        # use projection matrix K to do the mapping from 3D to 2D
+        points_2d = np.dot(K, point_in_camera_coords)
+
+        # normalize the x, y values by the third value
+        points_2d = np.array([
+            points_2d[0, :] / points_2d[2, :],
+            points_2d[1, :] / points_2d[2, :], points_2d[2, :]
+        ])
```
Comment (emailweixu) [inline alf/environments/carla_sensors.py]: Do we need the third dimension in the return value of points_2d?
```python
points_2d = points_2d[:2, :] / points_2d[2, :]
```

File: alf/environments/carla_sensors.py [hunk 2]
```diff
@@ -616,10 +618,124 @@ def render(self, display):
...
+        # transform the points from world space to camera space through the
+        # following several steps
+
+        # [N, 3] -> [3, N]
+        world_points = np.transpose(world_points, (1, 0))
+        # [3, N] -> [4, N]
+        world_points = np.concatenate(
+            (world_points, np.ones_like(world_points[0:1, ...])), axis=0)
+
+        sensor_points = np.dot(world_2_camera, world_points)
```
Comment (emailweixu) [inline alf/environments/carla_sensors.py]: is np.dot same as np.matmul here?

File: alf/environments/carla_sensors.py [hunk 3]
```diff
@@ -616,10 +618,124 @@ def render(self, display):
...
+
+        # change from UE4's left handed coordinate system to a typical
+        # right handed camera coordinate system, which is equivalent to
+        # axis swapping: (x, y, z) -> (y, -z, x)
+        point_in_camera_coords = np.array(
+            [sensor_points[1], sensor_points[2] * -1, sensor_points[0]])
```
Comment (emailweixu) [inline alf/environments/carla_sensors.py]: why not `-sensor_points[2]`?

File: alf/environments/carla_sensors.py [hunk 4]
```diff
@@ -616,10 +618,124 @@ def render(self, display):
...
+        points_2d = np.dot(K, point_in_camera_coords)
+
+        # normalize the x, y values by the third value
+        points_2d = np.array([
+            points_2d[0, :] / points_2d[2, :],
+            points_2d[1, :] / points_2d[2, :], points_2d[2, :]
+        ])
+
+        # [3, N] -> [N, 3]
+        points_2d = np.transpose(points_2d, (1, 0))
```
Comment (emailweixu) [inline alf/environments/carla_sensors.py]: `points_2d = points_2d.transpose()`

File: alf/environments/carla_sensors.py (hunk 1 above)
Comment (Haichao-Zhang) [reply alf/environments/carla_sensors.py]: Returning the third dimension is unnecessary. Updated according to suggestion.

File: alf/environments/carla_sensors.py (hunk 2 above)
Comment (Haichao-Zhang) [reply alf/environments/carla_sensors.py]: Yes they are the same as the arguments are 2D arrays. 
Changed ```np.dot``` to ```np.matmul```, as the latter one is preferred according to np documents for 2D case.

File: alf/environments/carla_sensors.py (hunk 3 above)
Comment (Haichao-Zhang) [reply alf/environments/carla_sensors.py]: Changed

**Review by emailweixu — APPROVED** [review]

**Pull request merged**